tools/sst_rewrite.cc

```diff
@@ -0,0 +1,353 @@
+    // disable these
+    LDbTable();
+    LDbTable(const LDbTable &);
+    const LDbTable operator=(const LDbTable&);
+};  // LDbTable
+
+
+LDbTable::LDbTable(
+    leveldb::Options & Options,
+    std::string & FileName)
+    : m_Options(Options), m_FileName(FileName),
+      m_FilePtr(NULL), m_TablePtr(NULL), m_FileSize(0), m_IsOpen(false)
+{
+    m_LastStatus=m_Options.env->GetFileSize(m_FileName, &m_FileSize);
+
+    if (m_LastStatus.ok())
+        m_LastStatus=m_Options.env->NewRandomAccessFile(m_FileName, &m_FilePtr);
+
+    if (m_LastStatus.ok())
+    {
+        m_LastStatus=leveldb::Table::Open(m_Options, m_FilePtr, m_FileSize, &m_TablePtr);
+
+        // use fadvise to start file pre-read
+        m_FilePtr->SetForCompaction(m_FileSize);
+    }   // if
+
+    m_IsOpen=m_LastStatus.ok();
+
+    if (!m_IsOpen)
+    {
+        // some people would throw() at this point, but not me
+        delete m_TablePtr;
+        m_TablePtr=NULL;
+        delete m_FilePtr;
+        m_FilePtr=NULL;
+        m_FileSize=0;
+    }   // if
+
+    return;
+
+}   // LDbTable::LDbTable
+
+
+LDbTable::~LDbTable()
+{
+    m_IsOpen=false;
+    delete m_TablePtr;
+    m_TablePtr=NULL;
+    delete m_FilePtr;
+    m_FilePtr=NULL;
+    m_FileSize=0;
+}   // LDbTable::~LDbTable
+
+
+leveldb::Iterator *
+LDbTable::NewIterator()
+{
+    leveldb::Iterator * ret_ptr(NULL);
+
+    if (m_IsOpen)
+    {
+        leveldb::ReadOptions read_options;
+
+        read_options.fill_cache=false;
+        ret_ptr=m_TablePtr->NewIterator(read_options);
+    }   // if
+
+    return(ret_ptr);
+
+}   // LDbTable::NewIterator
+
+
+int
+main(
+    int argc,
+    char ** argv)
+{
+    bool error_seen, running, compare_files;
+    char ** cursor;
+
+    compare_files=false;
+    error_seen=false;
+    running=true;
+
+    // Options: needs filter & total_leveldb_mem initialized
+    leveldb::Options options;
+    options.filter_policy=leveldb::NewBloomFilterPolicy2(16);
+    options.total_leveldb_mem=(512 << 20);
+
+    // testing
+//    options.block_size=16 << 10;
+
+    for (cursor=argv+1; NULL!=*cursor && running; ++cursor)
+    {
+        // option flag?
+        if ('-'==**cursor)
+        {
+            char flag;
+
+            flag=*((*cursor)+1);
+            switch(flag)
+            {
+                //case 'b':  options.block_size=4096; break;
+                case 'c':  compare_files=true; break;
+                case 'w':  compare_files=false; break;
+                default:
+                    fprintf(stderr, " option \'%c\' is not valid\n", flag);
+                    command_help();
+                    running=false;
+                    error_seen=true;
+                    break;
+            }   // switch
+        }   // if
+
+        // sst file
+        else
+        {
+            std::string fname;
+            fname=*cursor;
+
+            // do a rewrite
+            if (!compare_files)
+            {
+                leveldb::WritableFile *outfile = NULL;
```

Comment:
All of these objects that are allocated on the heap can be wrapped in std::auto_ptr<> objects and the explicit deletes below can be removed. If std::unique_ptr<> is available on all platforms, then that is preferred to auto_ptr, which is now deprecated.